src/reader.rs

```diff
@@ -72,18 +72,20 @@ impl<R: Read> BzDecompressor<R> {
 
 impl<R: Read> Read for BzDecompressor<R> {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        self.0.read(|stream, input, _eof| {
-            stream.decompress(input, buf)
+        self.0.read(buf.len(), |stream, input, offset, _eof| {
+            stream.decompress(input, &mut buf[offset..])
```

Comment:
I have yet to find a small, isolated test that demonstrates this was broken, but my app (a port of bsdiff, which uses a patch file format composed of three separate bzip2 streams) only works after this change.